### ISSUE TITLE
Bogus empty solutions with aggregates when no rows at all match

### ISSUE THREAD

**Issue by quasicomputational** [oxigraph/oxigraph#189]
This is with oxigraph 0.3.0-beta.3

Reproducer:

```
use oxigraph::store::Store;

const QUERY3: &str = "
SELECT (SAMPLE(?dx) as ?d)
WHERE
{
  ?c <https://prop.example/z> ?dx .
}
GROUP BY ?c
";

const QUERY4: &str = "
SELECT (?dx as ?d)
WHERE
{
  ?c <https://prop.example/z> ?dx .
}
";

const QUERY5: &str = "
SELECT (MAX(?dx) as ?d)
WHERE
{
  ?c <https://prop.example/z> ?dx .
}
GROUP BY ?c
";

#[tokio::main]
async fn main() -> Result<(), anyhow::Error> {
    let store = Store::new()?;

    match store.query(QUERY3)? {
        oxigraph::sparql::QueryResults::Solutions(solutions) => {
            println!("Query 3");
            let rows = solutions.collect::<Result<Vec<_>, _>>()?;
            for row in rows {
                println!("  Solution:");
                for (var, term) in row.iter() {
                    println!("    * {var} = {term}");
                }
            }
        },
        _ => panic!(),
    }

    match store.query(QUERY4)? {
        oxigraph::sparql::QueryResults::Solutions(solutions) => {
            println!("Query 4");
            let rows = solutions.collect::<Result<Vec<_>, _>>()?;
            for row in rows {
                println!("  Solution:");
                for (var, term) in row.iter() {
                    println!("    * {var} = {term}");
                }
            }
        },
        _ => panic!(),
    }

    match store.query(QUERY5)? {
        oxigraph::sparql::QueryResults::Solutions(solutions) => {
            println!("Query 5");
            let rows = solutions.collect::<Result<Vec<_>, _>>()?;
            for row in rows {
                println!("  Solution:");
                for (var, term) in row.iter() {
                    println!("    * {var} = {term}");
                }
            }
        },
        _ => panic!(),
    }

    Ok(())
}
```

Output:

```
Query 3
  Solution:
Query 4
Query 5
  Solution:
```

I haven't gone into the SPARQL spec to see if this is mandated & expected behaviour, but it's very surprising to me that the queries involving aggregates return a solution that has no variables in it. I would have expected no solutions at all to all three queries.

**Comment by Tpt** [comment]
Thank you so much for this so well written bug report!

SPARQL aggregates always return at least one group, even if the number of solutions is 0. This behavior is tested using [the http://www.w3.org/2009/sparql/docs/tests/data-sparql11/aggregates/manifest#agg-empty-group2 test from the the SPARQL 1.1 test suite](https://w3c.github.io/rdf-tests/sparql11/data-sparql11/aggregates/index.html#http://www.w3.org/2009/sparql/docs/tests/data-sparql11/aggregates/manifest#agg-empty-group2). 

This is indeed a strange behavior. I don't know what were the rationals for this. The definition of the [Group SPARQL operation](https://www.w3.org/TR/sparql11-query/#defn_algGroup) seems to suggest that there should be no solution. I chose to return an empty group in order to stay compatible with the official test suite.

**Comment by quasicomputational** [comment]
I probably should have googled before opening the issue, but apparently this is caused by the spec and the spec's test suite disagreeing (with a dash of the spec not covering the aggregates-but-no-grouping case properly). More context:

* https://afs.github.io/sparql-agg-group-empty
* eclipse/rdf4j#1978

I'm going to just hack around it with a `HAVING BOUND(...)` clause, which will do what I want it to; I don't know what the proper action is for oxigraph when the spec & its tests are confused, bearing in mind interopability risks on both sides of the ledger.

**Comment by afs** [comment]
The SPARQL tests are maintained by the community at https://github.com/w3c/rdf-tests/tree/main/sparql11/data-sparql11

It's not possible to change the working group test suites so w3c.org/2009/sparql/docs/tests/ should be considered to be the state at the time (they can't be changed without an active working group) and the github repo to be the current community consensus.

https://github.com/w3c/rdf-tests/pull/61 is fixes for the aggregate/group issues with community discussion/review/agreement.

**Comment by Tpt** [comment]
Thank you so much @afs for your comment. If the current consensus has moved to having no group if there is no solution and a `GROUP BY` let's do that too in Oxigraph (https://github.com/oxigraph/oxigraph/commit/d0b8fe1a23315b2eac0c2fe48fa65db6fa215a04)!

I am running Oxigraph tests using the latest version of w3c/rdf-tests. However, I discovered I made a mistake in the code checking the results. No solution was considered as similar to a single empty solution. I have fixed it in c2668ff881b0281830b0956b3cfd76f93a384ea1. It is why the new tests related to the number of groups where passing even if they should not. They should now pass thanks to d0b8fe1a23315b2eac0c2fe48fa65db6fa215a04.
